slave: monitor: add Associated STA Link Metrics Response

vitalii-komisarenko opened the pull request:
At the moment Associated STA Link Metrics are handled as
a part of *_HOSTAP_STATS_MEASUREMENT_* messages.
We would like to handle them separately.
    
This commit adds a flow for the Associated STA Link Metrics Response
that basically duplicates a subset of *_HOSTAP_STATS_MEASUREMENT_*.
It is a preparation for removing above-mentioned subset from Hostap Measurements.

vitalybu commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -179,11 +181,50 @@ void monitor_stats::process()
             sta_stats_msg.rx_rssi           = sta_stats.rx_rssi_curr;
+
+            beerocks_message::sBssidInfo bss_info;
+            bss_info.bssid                      = network_utils::mac_from_string(sta_mac);
```
`bssid` should be the MAC address of the AP not the STA.
Unless the field name is incorrect

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2088,6 +2095,56 @@ bool backhaul_manager::get_sta_bssid(const sMacAddr &sta_mac, sMacAddr &bssid)
+                                                                const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received ASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE , mid=" << std::dec
```
```suggestion
    LOG(DEBUG) << "Received ASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE, mid=" << std::dec
```

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2088,6 +2095,56 @@ bool backhaul_manager::get_sta_bssid(const sMacAddr &sta_mac, sMacAddr &bssid)
+        //TODO
+        return false;
+    }
+}
```
```suggestion
    }
    
    LOG(DEBUG) << "client with mac address " << mac->sta_mac() << " connected to " << bssid;
    //TODO
    return false;
}
```
`//TODO` what?

Also, if you already found the `bssid` to which the STA is connected, it's better to forward the CMDU the correct slave from here. Otherwise (by returning `false`) the message will be forwarded to all the slaves.

vitalybu commented on tests/test_flows.py:
```diff
@@ -488,6 +488,18 @@ def test_client_association_dummy(self):
         self.check_log(env.agents[0].radios[0],
                        r"Got client disallow request for {}".format(sta.mac))
 
+    def test_client_association_link_metrics(self):
```
Please this change to a separate commit

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2080,13 +2080,17 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
 
-bool backhaul_manager::get_sta_bssid(const sMacAddr &sta_mac, sMacAddr &bssid)
+bool backhaul_manager::get_sta_bssid_and_socket(const sMacAddr &sta_mac, sMacAddr &bssid,
+                                                Socket **socket)
```
It better to use a reference to a pointer (https://stackoverflow.com/a/3834103)
```suggestion
                                                Socket *&socket)
```

But why do you need to return both the `bssid` and the slave socket?
You can return the `sRadioInfo` to which the STA is connected and then extract both the `bssid` and the slave socket from there.

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2080,13 +2080,17 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
                 bssid = vap.first;
+                if (socket) {
+                    *socket = slave->slave;
+                }
```
Assuming a reference to a pointer:
```suggestion
                socket = slave->slave;
```

vitalybu commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1411,6 +1411,29 @@ bool slave_thread::handle_cmdu_backhaul_manager_message(
     }
+    case beerocks_message::ACTION_BACKHAUL_HOSTAP_STATS_MEASUREMENT_REQUEST: {
+        LOG(DEBUG) << "ACTION_BACKHAUL_HOSTAP_STATS_MEASUREMENT_REQUEST";
+        if (monitor_socket) {
```
According to our coding standards:
```suggestion
        if (!monitor_socket) {
            break;
        }
        ...        
```

vitalybu commented on tests/test_flows.py:
```diff
@@ -499,6 +499,19 @@ def test_client_association_link_metrics(self):
                                      tlv(0x95, 0x0006, '{sta_mac}'.format(sta_mac=sta_mac)))
         self.check_log(env.agents[0],
                        "client with mac address {sta_mac} not found".format(sta_mac=sta_mac))
+        time.sleep(1)
```
Again, tests should be updated in a separate commit

vitalii-komisarenko committed:
tlvf: enable wfa_map::tlvAssociatedStaLinkMetrics

Enable this tlv in order to be able process Associated STA Link Metrics queries.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
tlv: add ACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE

Added tlv able to transfer Associated STA Link Metrics from monitor to slave.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
slave: monitor: add Associated STA Link Metrics Response

At the moment Associated STA Link Metrics are handled as
a part of HOSTAP_STATS_MEASUREMENT messages.
We would like to handle them separately.

This commit adds a flow for the Associated STA Link Metrics Response
that basically duplicates a subset of HOSTAP_STATS_MEASUREMENT.
It is a preparation for removing above-mentioned subset from Hostap Measurements.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

Commit:
tlvf: add missing yamls for associated sta link metrics

Preparative commit.
Add tlvAssociatedStaLinkMetrics.yaml and tlvStaMacAddressType.yaml to
tlvf_conf.yaml, which are used when responding to an associated sta link
metrics query cmdu.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

Commit:
tlvf: update autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

Commit:
tlvf: add tlvStaMacAddressType to Cmdu parser

Preparative commit. Add tlvStaMacAddressType TLV to the parseNextTlv
method in CmduMessageRx so it will be automatically parsed on Rx.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
backhaul_manager: add get_sta_bssid helper

get_sta_bssid is a helper function to search associated client map to both:
* find bssid
* check if a sta is connected to given client map

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
backhaul_manager: add get_sta_radio helper

get_sta_radio was added to aid search in associated_client_map

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
backhaul_manager: add Associated STA Link Metrics error flow

Handling of error flow of Associated STA Link Metrics is added into backhaul manager.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
test_flows.py: Associated STA Link Metrics: error flow

When Associated STA Link Metrics Request is received with incorrect (not connected) STA mac,
The appropriated error tlv shall be attached to Associated STA Link Metrics Response.

This commit checks this flow.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
tlvf: add ACTION_BACKHAUL_HOSTAP_STATS_MEASUREMENT_REQUEST

This message is added in order to trigger HOSTAP_STATS_MEASUREMENT by backhaul.
It is needed for an intermediate implementation of Associated STA Link Metrics flow.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
agent: beerocks: propagate assocaited sta link metrics as hostap measurements

Hostap Measurements (HM) flow is being split into
Associated STA Link Metrics (ASLM) flow and the rest.

This commit is a step in this direction.
It processes ASLM request in backhaul, sends it to the son slave and then to monitor as a HM request.
Since as an intermediate step, HM response comes together with ASLM,
this commit makes it possible to process ASLM.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
test_flows: Associated STA Link Metrics: added test for connected STA

This commit adds test to check Associated STA Link Metrics to check the correct flow
(when STA with requested MAC address is present in the list of connected STAs)

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
cppcheck: add unfixable issues for associated sta link metrics

The implementation of Associated STA Link Metrics made cppcheck check fail.
There is not much sence to fix them, because either:
* they are errors in autogenerated code
* they existed before, but triggered the cppcheck only now

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2073,6 +2073,25 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+    const std::unordered_map<sMacAddr, associated_clients_t> &clients_map, const sMacAddr &sta_mac,
+    sMacAddr &bssid)
+{
+    return std::any_of(clients_map.begin(), clients_map.end(),
```
Can't it be done with ranged for-loops? They are more readable and have less indentation.
It is mention as preferred on https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#loops.

tomereli replied:
You are right Moran, but `cppcheck` thinks otherwise :/

arnout replied:
>> Can't it be done with ranged for-loops? They are more readable and have less indentation.

That they are more readable is a matter of taste.

>> It is mention as preferred on https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#loops.

That is only compared to "normal" for-loops, not compared to std::algorithms.


> You are right Moran, but `cppcheck` thinks otherwise :/

That's not good enough of a reason. If a tool doesn't do what we want, we have to change the tool, not what we want.

That said, I guess the main argument against using any_of is that it's too verbose? There are actually two ways that we can reduce the verbosity: by using C++14 auto, and by defining wrappers for the algorithms that take a container rather than an iterator.

```
    return any_of(clients_map, [&](const auto &vap) {
        return any_of(vap.second, [&](const auto &mac) {
            if (mac.first == sta_mac) {
                bssid = vap.first;
                return true;
            }
            return false;
        });
    });
```
where
```
template <class T, class UnaryPredicate>
bool any_of(const T& container, UnaryPredicate p)
{
    return any_of(std::begin(container), std::end(container), p);
}
```

Without that helper function, it's still not that bad if you use auto:

```
    return std::any_of(clients_map.begin(), clients_map.end(), [&](const auto &vap) {
        return std::any_of(vap.second.begin(), vap.second.end(), [&](const auto &mac) {
            if (mac.first == sta_mac) {
                bssid = vap.first;
                return true;
            }
            return false;
        });
    });
```

but I agree that it becomes bad when the lambda declaration no longer fits on the same line.

So I think the indentation argument is not really valid. Thus, the question is: which of the two is more readable?

I think we'll need to take a vote on #democracy on this.

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1411,6 +1411,29 @@ bool slave_thread::handle_cmdu_backhaul_manager_message(
+        if (monitor_socket) {
+            auto request_in = beerocks_header->addClass<
+                beerocks_message::cACTION_BACKHAUL_HOSTAP_STATS_MEASUREMENT_REQUEST>();
+            if (request_in == nullptr) {
```
With a smart pointer, prefer doing `if (!smart_ptr)` over `if (smart_ptr == nullptr)`

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1411,6 +1411,29 @@ bool slave_thread::handle_cmdu_backhaul_manager_message(
+                return false;
+            }
+            request_out->sync() = request_in->sync();
+            LOG(DEBUG) << "Will now send ACTION_MONITOR_HOSTAP_STATS_MEASUREMENT_REQUEST";
```
the "will now" is redundant.
`LOG(DEBUG) << "send ACTION_MONITOR_HOSTAP_STATS_MEASUREMENT_REQUEST";`

morantr reviewed (approved):
Nothing major so approving.
I really don't like the use of std::any instead of ranged loops.
also, lasts commit "cppcheck: existing issues: add unfixable issues for associated sta li……nk metrics" headline is too long.

tomereli commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
+            beerocks_message::sBssidInfo bss_info;
+            bss_info.bssid                      = network_utils::mac_from_string(sta_mac);
+            bss_info.earliest_measurement_delta = sta_stats.delta_ms;
+            // TODO: MAC data rate and Phy rate are not necessarily the same
```
TODO should be accompanied with a github issue URL

tomereli commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
+
+        auto sta_metrics = message_com::create_vs_message<
+            beerocks_message::cACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE>(cmdu_tx);
+        if (sta_metrics == nullptr) {
```
```suggestion
        if (!sta_metrics) {
```

tomereli commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
+        }
+
+        if (!sta_metrics->alloc_bssid_info_list(bss_infos.size())) {
+            LOG(ERROR) << "Failed allocate_bssid_info_list";
```
```suggestion
            LOG(ERROR) << "Failed allocate_bssid_info_list";
            return;
```

tomereli commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
+            LOG(ERROR) << "Failed allocate_bssid_info_list";
+        }
+
+        // TODO: set sta_metrics->sta_mac
```
TODO is both unclear and missing link to a github issue.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2073,6 +2073,25 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     }
 }
 
+bool backhaul_manager::get_sta_bssid(
```
Another option is to return sMacAddr instead of having it as a parameter - we can return an `network_utils::ZERO_MAC` if not found (nitpick).

vitalii-komisarenko replied:
shall `ZERO_MAC` convert implicitly to `false` in boolean context? :)

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2091,11 +2110,10 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
-                associated_client_found = true;
-                break;
-            }
+        sMacAddr bssid;
```
So this will change to:

```
auto bssid = get_sta_bssid(associated_clients_map, client_info_tlv_r->client_mac());
if (bssid != network_utils::ZERO_MAC) {
  associated_client_found = true;
  break;
}
```
(nitpick continued :) )

tomereli reviewed (commented):
Not finished.

vitalii-komisarenko committed:
test_flows: test Associated Link Metrics Response of backhaul

Associated STA Link Metrics Response flow was changed to go through
backhaul manager, because it can set `bssid` field.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
test_flows: Associated STA Link Metrics: check message ID

Associated STA Link Metrics request and response shall have the same message ID (MID).
test_flow is edited to check that MID of the response corresponds to MID of the request.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
tlvf: add MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE

Added ACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE.
Associated STA Link Metrics need to go through backhaul_manager
in order to have `bssid` field set.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
beerocks: Associated Link Metrics: fill BSSID in backhaul

The Associated STA Link Metrics Response is being filled in monitor thread,
then sent to agent and then to controller.
This flow is imperfect because `BSSID` field is not set.

This commit changes the flow, so that the message "visits" the backhaul manager
between the agent and the controller in order to have BSSID set there.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
agent: monitor_stats: add MAC address into request

Monitor_stats requests were designed to serve Hostap Measurements (HM) requests.
We want to adapt the existing code to support Associated STA Link Measurements (ASLM) requests.

One of the differences between these two reqeusts in that
HM requests require information about all the STAs,
but ASLM ones need only data about a specific one.

This commit adds a MAC address field into request in order to differentiate them,
ZERO_MAC (00:00:00:00:00:00) meaning HM request,
all other -- ASLM for the STA with this specific MAC address.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
beerocks: rename BACKHAUL_HOSTAP_MEASUREMENTS

Rename BACKHAUL_HOSTAP_MEASUREMENTS_REQUEST into BACKHAUL_ASSOCIATED_STA_LINK_METRICS_REQUEST
in order to better reflect its purpose.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
tlvf: add ACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_REQUEST

Currently Hostap Measurements request is used instead of
ACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_REQUEST,
but it lacks MAC address field.
Associated Link Measurements require MAC field because
it collects information about one specific STA.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
tlvf: add sta_mac to ACTION_BACKHAUL_ASSOCIATED_STA_LINK_METRICS_REQUEST

Associated STA Link Metrics Request needs MAC address in order to
specify the STA.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
beerocks: Associated Link Metrics: add mac address

Send `sta_mac` field from backhaul manager to son_slave
and then to monitor_tread in order to make monitoring request specific to a single STA.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
monitor_stats: associated sta link metrics request support

Depending on the value of MAC address of the request treat it as
a hostap metrics request (MAC=00:00:00:00:00:00) or associated sta link metrics (otherwise).

Both requests procession are also moved to separate functions.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
monitor_stats: associated link metrics: turn `for` into `find_if`

To serve Associated STA Link Metrics flow, the monitor_stats looks through
a list of STA in order to find the correct one.

According to our current coding standard, std algorithms are preferred over `for` loops,
so the main loop of this search is turned into `std::find_if`.

It also has a nice side effect: it is trivial to check if the desired STA has been found.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

vitalii-komisarenko committed:
monitor_stats: add debug output before serving a request

The old request flow in monitor_stats has been extended to support both
Associated STA Link Metrics and Hostap Stats Measurements.

At the moment Associated STA Link Metrics requests are working in `test_flows.py`,
but are lost during certification tests.

The debug output before serving both types of requests is added in order to
investigate this issue further.

Signed-off-by: Vitalii Komisarenko <vitalii.komisarenko@gmail.com>

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -976,3 +976,16 @@ sChannelScanResults:
     _comment: Indicates the fraction of the time AP senses that the channel is in use by the neighboring AP for transmissions.
     _type: uint32_t
+
+sBssidInfo:
```
(nitpick) I don't think the name is very good. It's not info about a BSS, it's the STA link metrics for one particular BSS... Maybe `sSingleSTALinkMetric`?

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_monitor.yaml:
```diff
@@ -90,6 +90,18 @@ cACTION_MONITOR_HOSTAP_ACTIVITY_NOTIFICATION:
   _type: class
   params: sApActivityNotificationParams 
 
+cACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE:
```
In #1198 we decided to use the IEEE1905.1 message type directly rather than defining a new TLV that copies everything from the EasyMesh message. You could do the same here...

That said, it's probably a lot of work to rework this big PR to do that now. So I'm not going to say you have to do it.

arnout commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -220,11 +222,65 @@ void monitor_stats::process()
+                auto sta_metrics = message_com::create_vs_message<
+                    beerocks_message::cACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE>(
+                    cmdu_tx, message_id);
+                if (sta_metrics == nullptr) {
```
```suggestion
                if (!sta_metrics) {
```
https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#klocwork

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2146,15 +2156,7 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
-            break;
-        }
-    }
+    bool associated_client_found = (bool)get_sta_radio(client_info_tlv_r->client_mac());
```
Please no C-style casts. I believe this works:
```suggestion
    bool associated_client_found { get_sta_radio(client_info_tlv_r->client_mac()) };
```

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
+                                                                const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received ASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE , mid=" << std::dec
```
I believe the std::dec shouldn't be needed any more because we reset the formatters in the LOG destructor, no?

mariomaz replied:
Most of the times the mid value is formatted as hex with `std::hex` while logging

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
+        return false;
+    }
+
+    auto mac = cmdu_rx.getClass<wfa_map::tlvStaMacAddressType>();
```
(nitpick) mac can be anything, sta_mac would be better.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
+    }
+
+    auto mac = cmdu_rx.getClass<wfa_map::tlvStaMacAddressType>();
+    if (!mac) {
```
(nitpick) do all checks before the cmdu_tx::create.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
+    // with any of the BSS operated by the Multi-AP Agent
+    std::shared_ptr<sRadioInfo> radio = get_sta_radio(mac->sta_mac());
+    if (!radio) {
+        LOG(ERROR) << "client with mac address " << mac->sta_mac() << " not found";
```
This is not really an error, rather info or warning. It is expected to happen, occasionally, because the controller may send the request just as the station is disassociating.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
+        }
+        LOG(DEBUG) << "client with mac address " << mac->sta_mac() << " connected to " << bssid;
+        //TODO: non-error flow for Associated STA Link Metrics Request
+        return false;
```
(nitpick) since you already created the response message, I'd still send it here but without the required TLV. But since you fix that later in the same PR it's not important.

arnout commented on tests/test_flows.py:
```diff
@@ -531,6 +531,18 @@ def test_client_association_dummy(self):
+        an Associated STA Link Metrics Query message with an Associated STA Link Metrics
+        Response message containing an Associated STA Link Metrics TLV for the associated STA.'''
+
+        sta_mac = "11:11:33:44:55:66"
```
```suggestion
        sta = env.Station.create()
```

arnout commented on tests/test_flows.py:
```diff
@@ -531,6 +531,18 @@ def test_client_association_dummy(self):
+        sta_mac = "11:11:33:44:55:66"
+        debug("Send link metrics query for unconnected STA")
+        env.controller.dev_send_1905(env.agents[0].mac, 0x800D,
+                                     tlv(0x95, 0x0006, '{sta_mac}'.format(sta_mac=sta_mac)))
```
```suggestion
                                     tlv(0x95, 0x0006, '{}'.format(sta.mac)))
```
The `{var}` form is very useful in f-strings, but without them it's just too verbose IMO.

arnout commented on tests/test_flows.py:
```diff
@@ -531,6 +531,18 @@ def test_client_association_dummy(self):
+        debug("Send link metrics query for unconnected STA")
+        env.controller.dev_send_1905(env.agents[0].mac, 0x800D,
+                                     tlv(0x95, 0x0006, '{sta_mac}'.format(sta_mac=sta_mac)))
+        self.check_log(env.agents[0],
```
We should actually check the error TLV in the reply, but that requires #1142 of course...

arnout commented on tests/test_flows.py:
```diff
@@ -542,6 +542,19 @@ def test_client_association_link_metrics(self):
                        "client with mac address {sta_mac} not found".format(sta_mac=sta_mac))
+        time.sleep(1)
+
+        sta = env.Station.create()
```
To distinguish from the previous sta, use sta1 and sta2.

arnout commented on ci/cppcheck/cppcheck_existing_issues.txt:
```diff
@@ -678,3 +678,7 @@
 '/builds/prpl-foundation/prplMesh/controller/src/beerocks/master/son_management.cpp: style: Variable 'op_error_code' is assigned a value that is never used. [unreadVariable]        auto op_error_code = eChannelScanOpErrCode::CHANNEL_SCAN_OP_SUCCESS;
 '/builds/prpl-foundation/prplMesh/controller/src/beerocks/master/son_management.cpp: style: Variable 'op_error_code' is assigned a value that is never used. [unreadVariable]            op_error_code = eChannelScanOpErrCode::CHANNEL_SCAN_OP_SCAN_IN_PROGRESS;
 '/builds/prpl-foundation/prplMesh/controller/src/beerocks/master/son_management.cpp: style: Variable 'result_status' is assigned a value that is never used. [unreadVariable]            result_status = last_scan_success;
+'/builds/prpl-foundation/prplMesh/agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h: warning: Virtual function 'on_thread_stop' is called from destructor '~backhaul_manager()' at line 332. Dynamic binding is not used. [virtualCallInConstructor]    virtual void on_thread_stop() override;
```
This one was already in there, it shouldn't be duplicated.

To regenerate this file, just run ci/cppcheck/cppcheck-new-issues.sh manually and copy cppcheck_results.txt to this file. Or better yet let CI do the job and take it from the cppcheck-diff artefacts.

arnout commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -220,11 +222,65 @@ void monitor_stats::process()
 
             sta_count++;
+
+            beerocks_message::sBssidInfo bss_info;
```
You should also get bssid here. It's a bit hard to get, because the sta_node only has a vap_id:
```
    bss_info.bssid = network_utils::mac_from_string(mon_db->vap_get_by_id(sta_node->get_vap_id())->get_mac());
```
and you need to handle the possible nullpointer from vap_get_by_id.

But setting bssid this way removes the need for the backhaul manager to be involved in the response path.

arnout commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -51,7 +51,7 @@ void monitor_stats::add_request(uint16_t id, uint8_t sync)
         next_poll_id = mon_db->get_poll_id() + 1;
     }
-    requests_list.push_back(id);
+    requests_list.push_back(sMeasurementsRequest(id, sta_mac));
```
```suggestion
    requests_list.push_back({ id, sta_mac });
```
With that, the constructor is not needed.

arnout commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```suggestion
        if (!sta_node) {
```

arnout reviewed (approved):
Although I have a lot of nitpick comments, there is only one that is important enough to be absolutely required to change: don't use C-style cast.

Feel free to ignore all the other comments.

I propose to merge this PR even though it doesn't pass yet in certification.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2142,6 +2142,16 @@ sMacAddr backhaul_manager::get_sta_bssid(
+backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
+{
+    auto radio = std::find_if(
+        slaves_sockets.begin(), slaves_sockets.end(), [&sta_mac](std::shared_ptr<sRadioInfo> r) {
```
I think that parameter to predicate function should be const reference
```
const std::shared_ptr<sRadioInfo> &radio_info
```

rmelotte replied:
Good point. 
Also, since it does not share the ownership of the sRadioInfo, I don't see why it couldn't be a `sRadioInfo &` instead of a shared_ptr.

rmelotte reviewed (approved):
:+1: 
There are still some nitpick comments remaining, but as this PR is needed for https://github.com/prplfoundation/prplMesh/issues/394 and is a giant leap forward for 4.7.7, let's merge it!
